### PR TITLE
Availability: Adds improved cache logic for partition key ranges

### PR DESCRIPTION
**Pull request by j82w**
## Description

This is converting the partition key range cache to use the new AsyncCacheNonBlocking. 

Issues being fixed:
1. The current implementation is once a refresh occurs all other requests are blocked until the refresh completes. Even if they have 1000 partitions and only 1 is splitting all 999 partitions will be blocked until the refresh completes.
2. If the refresh fails the old cache value is lost. Then the SDK has to reload the entire which cause additional load to the service on the meta data requests.
3. Fixes a bug in the AsyncCacheNonBlocking cache were if a request is not awaited on the initial request the value could get stuck returning the original task.
4. Fixes a bug where if a 404 error is throne the stale value is not removed from the cache.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes https://github.com/Azure/azure-cosmos-dotnet-v3/issues/3060

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 1]
```diff
@@ -915,7 +915,7 @@ private async Task OpenPrivateAsync(CancellationToken cancellationToken)
             // For direct: WFStoreProxy [set in OpenAsync()].
             this.eventSource = DocumentClientEventSource.Instance;
 
-            this.initializeTaskFactory = () =>
+            this.initializeTaskFactory = (staleValue) =>
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs:918]: is the stale value used anywhere?

File: Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs [hunk 2]
```diff
@@ -319,7 +329,6 @@ private void ThrowIfTargetChanged(DocumentServiceRequest request, PartitionKeyRa
                 if (!collectionCacheIsUptoDate)
                 {
                     request.ForceNameCacheRefresh = true;
-                    collectionCacheIsUptoDate = true;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs:322]: When is this set now?

File: Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs [hunk 3]
```diff
@@ -68,27 +83,65 @@ public AsyncCacheNonBlocking()
...
+                        return await initialLazyValue.GetValueAsync();
+                    }
+                    catch (Exception e)
+                    {
+                        // This is needed for scenarios where the initial GetAsync was
+                        // called but never awaited.
+                        if (initialLazyValue.ShouldRemoveFromCacheThreadSafe())
+                        {
+                            DefaultTrace.TraceError(
+                                "AsyncCacheNonBlocking Failed GetAsync. Init task was not removed with key: {0}, Exception: {1}",
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs:105]: It says was not removed but after it does a TryRemove?

File: Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs [hunk 4]
```diff
@@ -191,7 +191,7 @@ internal class GatewayAddressCache : IAddressCache, IDisposable
                 {
                     addresses = await this.serverPartitionAddressCache.GetAsync(
                         key: partitionKeyRangeIdentity,
-                        singleValueInitFunc: () => this.GetAddressesForRangeIdAsync(
+                        singleValueInitFunc: (stale) => this.GetAddressesForRangeIdAsync(
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs:194]: Should we just use `_` since the value is not used?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs:194]: Good point, I'll update it.

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/DocumentClient.cs:918]: It is used in both the PK Range and Address caches. Address cache is used to log if there was any changes and what those changes are. The PK range cache uses it for the changefeed scenario so it only updates based on the latest version.

File: Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs:322]: This fixes a VS warning. Nothing else uses the flag. There is no point in setting it.

File: Microsoft.Azure.Cosmos/src/Query/v2Query/PartitionKeyRangeGoneRetryPolicy.cs [hunk 5]
```diff
@@ -121,17 +121,11 @@ public void OnBeforeSendRequest(DocumentServiceRequest request)
...
-                    {
-                        // Force refresh.
-                        await this.partitionKeyRangeCache.TryLookupAsync(
-                                collection.ResourceId,
-                                routingMap,
-                                request,
-                                cancellationToken,
-                                this.trace);
-                    }
+                    CollectionRoutingMap routingMap = await this.partitionKeyRangeCache.TryLookupAsync(
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Query/v2Query/PartitionKeyRangeGoneRetryPolicy.cs:124]: @j82w one more instance where we are managing this dead code.

File: Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs [hunk 6]
```diff
@@ -328,22 +337,19 @@ private void ThrowIfTargetChanged(DocumentServiceRequest request, PartitionKeyRa
                         collectionRoutingMapCacheIsUptoDate = false;
                         routingMap = await this.collectionRoutingMapCache.TryLookupAsync(
                             collection.ResourceId,
-                            previousValue: null,
                             request: request,
-                            cancellationToken: cancellationToken,
-                            trace: NoOpTrace.Singleton);
+                            trace: NoOpTrace.Singleton,
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs:341]: we are no more passing cancellation token in lookup calls. does it mean...these lookups cannot be cancelled once started?

File: Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs [hunk 7]
```diff
@@ -237,7 +297,15 @@ public Task<T> GetValueAsync()
                     else
                     {
                         createdTask = true;
-                        this.refreshInProgress = createRefreshTask();
+                        T previousValue = default;
+                        if (this.value.IsCompleted && !this.value.IsFaulted && !this.value.IsCanceled)
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs:301]: nit: IsCanceled spelling is wrong.

File: Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs [hunk 8]
```diff
@@ -191,7 +191,7 @@ internal class GatewayAddressCache : IAddressCache, IDisposable
                 {
                     addresses = await this.serverPartitionAddressCache.GetAsync(
                         key: partitionKeyRangeIdentity,
-                        singleValueInitFunc: () => this.GetAddressesForRangeIdAsync(
+                        singleValueInitFunc: (_) => this.GetAddressesForRangeIdAsync(
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs:194]: is it something visual studio suggested?

File: Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs (hunk 6 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs:341]: The caches should never be cancelled by a single request because there is likely multiple other requests waiting for that update.

File: Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs (hunk 7 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs:301]: That is part of .NET. This can not be fixed: https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.iscanceled?view=net-6.0

File: Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs (hunk 8 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs:194]: This was required because it now passes the stale value into the delegate.

File: Microsoft.Azure.Cosmos/src/Query/v2Query/PartitionKeyRangeGoneRetryPolicy.cs (hunk 5 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/v2Query/PartitionKeyRangeGoneRetryPolicy.cs:124]: Correct, but this owned by the query team and should be handled in a separate PR.

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs (hunk 1 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/DocumentClient.cs:918]: But there is no reference to this variable within the scope, my point is, should we use `_`?

File: Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs [hunk 9]
```diff
@@ -263,15 +263,22 @@ private void ThrowIfTargetChanged(DocumentServiceRequest request, PartitionKeyRa
 
             ContainerProperties collection = await this.collectionCache.ResolveCollectionAsync(request, cancellationToken, NoOpTrace.Singleton);
             CollectionRoutingMap routingMap = await this.collectionRoutingMapCache.TryLookupAsync(
-                collection.ResourceId, null, request, cancellationToken, NoOpTrace.Singleton);
+                collection.ResourceId,
+                request,
+                NoOpTrace.Singleton,
+                false);
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/Routing/AddressResolver.cs:269]: NIT: named parameter please

**Review by FabianMeiswinkel — DISMISSED** [review]
Nice - well designed!

File: Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs [hunk 10]
```diff
@@ -24,24 +25,38 @@ internal sealed class AsyncCacheNonBlocking<TKey, TValue> : IDisposable
...
 
         public AsyncCacheNonBlocking()
-            : this(valueEqualityComparer: EqualityComparer<TValue>.Default, keyEqualityComparer: null)
+            : this(keyEqualityComparer: null)
         {
         }
 
+        public static bool RemoveNotFoundFromCacheOnException(Exception e)
+        {
+            if (e is DocumentClientException dce
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs:45]: nit: This sounds very specific to cache calls that interact with network, and this is more like a generic cache implementation? Just wondering if it should be somewhere else, but nothing blocking

File: Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs [hunk 11]
```diff
@@ -139,18 +137,23 @@ internal class PartitionKeyRangeCache : IRoutingMapProvider, ICollectionRoutingM
...
+            IClientSideRequestStatistics clientSideRequestStatistics)
         {
             try
             {
                 CollectionRoutingMap routingMap = await this.routingMapCache.GetAsync(
                     collectionRid,
-                    null,
-                    () => this.GetRoutingMapForCollectionAsync(collectionRid, null, trace, clientSideRequestStatistics, CancellationToken.None),
-                    CancellationToken.None);
+                    (previousValue) => this.GetRoutingMapForCollectionAsync(
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs:150]: This async operation has a cancellationtoken that we were previously passing as None, but is there a chance to wire it down? `TryLookupAsync` used to receive it.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs:150]: Or maybe we were just passing None from up top

**Review by ealsur — DISMISSED** [review]
LGTM, left some non-blocking comments in case you think can/should be addressed

File: Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs (hunk 11 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs:150]: The cancellation token should not be passed into the cache, because in most scenarios multiple requests are waiting on that cache refresh. If the cancellation token fails the refresh it will impact all the other tasks waiting on it. It's better to let the cache operation complete and then cancel the request before the next network operation.

File: Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs (hunk 10 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs:48]: Added delegate that allows creator to override it.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 12]
```diff
@@ -367,20 +367,9 @@ public void Dispose()
...
-                {
-                    collectionRoutingMap = await partitionKeyRangeCache.TryLookupAsync(
-                        collectionRid: collection.ResourceId,
-                        previousValue: collectionRoutingMap,
-                        request: request,
-                        cancellationToken: CancellationToken.None,
-                        NoOpTrace.Singleton);
-                }
+                    trace: NoOpTrace.Singleton,
+                    forceRefresh: refreshCache);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:372]: Nice abstraction. Aoid this code duplication/spilling

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:372]: Typical forceRefresh is to ensure that a newer value than requested one.
But above semantics are different right?

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 13]
```diff
@@ -915,7 +915,7 @@ private async Task OpenPrivateAsync(CancellationToken cancellationToken)
             // For direct: WFStoreProxy [set in OpenAsync()].
             this.eventSource = DocumentClientEventSource.Instance;
 
-            this.initializeTaskFactory = () =>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs:918]: Its not clear on why this change is needed for this PR?

Its private and implementation detail for documentclient and none other of DC are changing.

File: Microsoft.Azure.Cosmos/src/Query/v2Query/PartitionKeyRangeGoneRetryPolicy.cs [hunk 14]
```diff
@@ -121,17 +121,11 @@ public void OnBeforeSendRequest(DocumentServiceRequest request)
...
+                        collectionRid: collection.ResourceId,
+                        request: request,
+                        forceRefresh: true,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/v2Query/PartitionKeyRangeGoneRetryPolicy.cs:127]: Semantically its different from our regular forceRefresh right?

File: Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs [hunk 15]
```diff
@@ -36,9 +36,8 @@ internal class PartitionKeyRangeCache : IRoutingMapProvider, ICollectionRoutingM
             IStoreModel storeModel,
             CollectionCache collectionCache)
         {
-            this.routingMapCache = new AsyncCache<string, CollectionRoutingMap>(
-                    EqualityComparer<CollectionRoutingMap>.Default,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs:40]: Equality comparer is missing. Does it matter?

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs (hunk 13 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/DocumentClient.cs:918]: This is needed in the partition key range cache so the update always gets the latest cache value to avoid doing requests with older values.

File: Microsoft.Azure.Cosmos/src/Query/v2Query/PartitionKeyRangeGoneRetryPolicy.cs (hunk 14 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/v2Query/PartitionKeyRangeGoneRetryPolicy.cs:127]: How is this semantically different from the regular forceRefresh? It always uses a bool flag to tell the cache to do a force refresh.

File: Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs (hunk 15 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/PartitionKeyRangeCache.cs:40]: The equality comparer is for the value. There is no point is comparing the values.
